kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/arbs.kt

```diff
@@ -64,28 +64,48 @@ fun <A, B> Arb<A>.flatMap(f: (A) -> Arb<B>): Arb<B> = object : Arb<B>() {
 }
 
 /**
- * Returns a new [Arb] which ensures only unique values are generated by keeping track of previously
- * generated values.
+ * Returns a new [Arb] which ensures samples are drawn from uniformly distributed distinct elements from
+ * the original distribution.
  *
- * Note: This arb can result in an infinite loop if more elements are requested than can be generated uniquely.
+ * The number of minimum sample size drawn randomly from the original distribution is determined by [minSampleSize].
+ * Since resampling can be very expensive, the randomly sampled elements from the original distribution will be
+ * cached unless [cacheSamples] is set to false.
  */
-fun <A> Arb<A>.distinct() = distinctBy { it }
+fun <A> Arb<A>.distinct(minSampleSize: Int = 1000, cacheSamples: Boolean = true): Arb<A> =
+   distinctBy(minSampleSize, cacheSamples) { it }
 
 /**
- * Returns a new [Arb] which ensures only distinct keys returned by the given [selector] function are generated by
- * keeping track of previously generated values.
+ * Returns a new [Arb] which ensures samples are drawn from uniformly distributed distinct elements from
+ * the original distribution based on the provided [selector] function.
  *
- * Note: This arb can result in an infinite loop if more elements are requested than can be generated uniquely.
+ * The number of minimum sample size drawn randomly from the original distribution is determined by [minSampleSize].
+ * Since resampling can be very expensive, the randomly sampled elements from the original distribution will be
+ * cached unless [cacheSamples] is set to false.
  */
-fun <A, B> Arb<A>.distinctBy(selector: (A) -> B) = object : Arb<A>() {
+fun <A, B> Arb<A>.distinctBy(
+   minSampleSize: Int = 1000,
+   cacheSamples: Boolean = true,
+   selector: (A) -> B
+): Arb<A> = object : Arb<A>() {
    override fun edgecases(): List<A> = this@distinctBy.edgecases().distinctBy(selector)
-
-   override fun values(rs: RandomSource): Sequence<Sample<A>> {
-      return this@distinctBy.values(rs).distinctBy { selector(it.value) }
+   override fun sample(rs: RandomSource): Sample<A> {
+      val sampledPopulation = if (cacheSamples) {
+         if (sampledElements.size < minSampleSize) {
+            val sampled = sampleFromPopulation(rs)
+            sampledElements.addAll(sampled)
+            sampled
+         } else sampledElements
+      } else sampleFromPopulation(rs)
+
+      return Sample(sampledPopulation.distinctBy(selector).random(rs.random))
    }
 
-   override fun sample(rs: RandomSource): Sample<A> =
-      samples(rs).distinctBy { selector(it.value) }.first()
+   override fun values(rs: RandomSource): Sequence<Sample<A>> = generateSequence { sample(rs) }
+
+   private fun sampleFromPopulation(rs: RandomSource): List<A> =
+      this@distinctBy.generate(rs).map { it.value }.take(minSampleSize).toList()
+
+   private val sampledElements: MutableList<A> = mutableListOf()
```

Comment:
I don't think this is a good idea because it makes tests less repeatable. Another approach would be not to cache it or to always generate this list with a static RandomSeed. Summoning @sksamuel for guidance

Reply:
I don't think this addresses the underlying concern, that this function can hang because you may never get enough random values.

One way around that is to take a parameter which is how many times you'll attempt to get the next element say `maxUniqueAttempts` or something.

Reply:
In terms of hanging - it should not really hang anymore but i'm not entirely convinced of what we're trying to solve here. This function looks more like a `Arb.resampleToUniformRandom()` at this point which is essentially sampling from an underlying distinct distribution. i.e. `Arb.of(1, 2, 3 ,4, 5, 6, 7, 8)`

Reply:
I've marked this function as deprecated. The primary problem with `distinct` is that it is terminal and stateful. Hence in terms of actually implementing the behaviour that we are after, i.e. sampling without replacement, we need to rethink the return type. Arb<A> is not capable of modelling this, as it is a population, it is infinite, and stateless. We may need to think of a type that is terminal. `Exhaustive<A>` is backed by `List<A>` under the hood so that might be an option. However, as you've rightly mention, Exhaustive itself is probably not suitable either. There's nothing in its type that says generate(rs) will emit only unique elements without replacement and terminates.